### PR TITLE
[INLONG-7635][Sort] MongoDB CDC supports metrics with enable incremental snapshots

### LINKED ISSUE #7635: [Feature][Sort] MongoDB CDC supports metrics with enable incremental snapshots
### Description

MongoDB CDC supports metrics with enable incremental snapshots

### Use case

none.

### Are you willing to submit PR?

- [X] Yes, I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by e-mhui**
### Prepare a Pull Request

MongoDB CDC supports metrics with enable incremental snapshots

- Fixes #7635 

### Motivation

MongoDB CDC supports metrics with enable incremental snapshots.

### Modifications

To be added later.

1. Register metrics.

position: `inlong-sort/sort-connectors/mongodb-cdc/src/main/java/org/apache/inlong/sort/cdc/base/source/IncrementalSource.java`

```java

    @Override
    public IncrementalSourceReader<T, C> createReader(SourceReaderContext readerContext)
            throws Exception {
        // create source config for the given subtask (e.g. unique server id)
        C sourceConfig = configFactory.create(readerContext.getIndexOfSubtask());
        MetricConfig metricConfig = (MetricConfig) sourceConfig;
        FutureCompletingBlockingQueue<RecordsWithSplitIds<SourceRecords>> elementsQueue =
                new FutureCompletingBlockingQueue<>();

        // Forward compatible with flink 1.13
        final Method metricGroupMethod = readerContext.getClass().getMethod("metricGroup");
        metricGroupMethod.setAccessible(true);
        final MetricGroup metricGroup = (MetricGroup) metricGroupMethod.invoke(readerContext);
        final SourceReaderMetrics sourceReaderMetrics = new SourceReaderMetrics(metricGroup);

        // create source config for the given subtask (e.g. unique server id)
        MetricOption metricOption = MetricOption.builder()
                .withInlongLabels(metricConfig.getInlongMetric())
                .withAuditAddress(metricConfig.getInlongAudit())
                .withRegisterMetric(RegisteredMetric.ALL)
                .build();

        sourceReaderMetrics.registerMetrics(metricOption,
                Arrays.asList(Constants.DATABASE_NAME, Constants.COLLECTION_NAME));
       /** **/
    }


```

2. Collect metrics.

position: `inlong-sort/sort-connectors/mongodb-cdc/src/main/java/org/apache/inlong/sort/cdc/mongodb/source/reader/MongoDBRecordEmitter.java`

```java
   @Override
    protected void processElement(
            SourceRecord element, SourceOutput<T> output, SourceSplitState splitState)
            throws Exception {
        if (isWatermarkEvent(element)) {
            Offset watermark = getOffsetPosition(element);
            if (isHighWatermarkEvent(element) && splitState.isSnapshotSplitState()) {
                splitState.asSnapshotSplitState().setHighWatermark(watermark);
            }
        } else if (isHeartbeatEvent(element)) {
            if (splitState.isStreamSplitState()) {
                updatePositionForStreamSplit(element, splitState);
            }
        } else if (isDataChangeRecord(element)) {
            if (splitState.isStreamSplitState()) {
                updatePositionForStreamSplit(element, splitState);
            }
            reportMetrics(element);
            debeziumDeserializationSchema.deserialize(
                    element,
                    new Collector<T>() {

                        @Override
                        public void collect(final T t) {
                            Struct value = (Struct) element.value();
                            Struct source = value.getStruct(MongoDBEnvelope.NAMESPACE_FIELD);
                            if (null == source) {
                                source = value.getStruct(RecordUtils.DOCUMENT_TO_FIELD);
                            }
                            String dbName = source.getString(MongoDBEnvelope.NAMESPACE_DATABASE_FIELD);
                            String collectionName =
                                    source.getString(MongoDBEnvelope.NAMESPACE_COLLECTION_FIELD);
                            sourceReaderMetrics
                                    .outputMetrics(dbName, collectionName, splitState.isSnapshotSplitState(), value);
                            output.collect(t);
                        }

                        @Override
                        public void close() {
                            // do nothing
                        }
                    });
        } else {
            // unknown element
            LOG.info("Meet unknown element {}, just skip.", element);
        }
    }
```

3. Restore metrics.

position: `inlong-sort/sort-connectors/cdc-base/src/main/java/org/apache/inlong/sort/cdc/base/source/reader/IncrementalSourceReader.java`

```java

    @Override
    public void addSplits(List<SourceSplitBase> splits) {
        // restore for finishedUnackedSplits
        List<SourceSplitBase> unfinishedSplits = new ArrayList<>();
        for (SourceSplitBase split : splits) {
            if (split instanceof MetricSplit) {
                MetricSplit metricSplit = (MetricSplit) split;
                LOG.info("inlong-metric-states restore metricSplit:{}", metricSplit);
                sourceReaderMetrics.initMetrics(metricSplit.getNumRecordsIn(),
                        metricSplit.getNumBytesIn(), metricSplit.getReadPhaseMetricMap(),
                        metricSplit.getTableMetricMap());
                LOG.info("inlong-metric-states restore sourceReaderMetrics:{}",
                        sourceReaderMetrics.getSourceMetricData());
                continue;
            }
            if (split.isSnapshotSplit()) {
                SnapshotSplit snapshotSplit = split.asSnapshotSplit();
                if (snapshotSplit.isSnapshotReadFinished()) {
                    finishedUnackedSplits.put(snapshotSplit.splitId(), snapshotSplit);
                } else {
                    unfinishedSplits.add(split);
                }
            } else {
                // the stream split is uncompleted
                if (!split.asStreamSplit().isCompletedSplit()) {
                    uncompletedStreamSplits.put(split.splitId(), split.asStreamSplit());
                    requestStreamSplitMetaIfNeeded(split.asStreamSplit());
                } else {
                    uncompletedStreamSplits.remove(split.splitId());
                    StreamSplit streamSplit =
                            discoverTableSchemasForStreamSplit(split.asStreamSplit());
                    unfinishedSplits.add(streamSplit);
                }
            }
        }
        // notify split enumerator again about the finished unacked snapshot splits
        reportFinishedSnapshotSplitsIfNeed();
        // add all un-finished splits (including stream split) to SourceReaderBase
        super.addSplits(unfinishedSplits);
    }

```

### CODE REVIEW

**Comment by e-mhui** [comment]
@gong @yunqingmoswu @EMsnap @kuansix Hi, pls review it.

**Comment by kuansix** [comment]
来信已收到，谢谢！

**Comment by e-mhui** [comment]
> I suggest mongo module don't use same package path with cdc-base module

Thanks, pls review this pr https://github.com/apache/inlong/pull/7640 .

**Review by gong — APPROVED** [review]

**Review by yunqingmoswu — APPROVED** [review]

**Review by EMsnap — APPROVED** [review]

File: inlong-sort/sort-connectors/mongodb-cdc/src/main/java/org/apache/inlong/sort/cdc/mongodb/source/utils/BsonUtils.java [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+import org.bson.BsonString;
+import org.bson.BsonType;
+import org.bson.BsonUndefined;
+import org.bson.BsonValue;
+import org.bson.types.Decimal128;
+
+/** Utility class to deal {@link BsonValue}s.
+ * Copy from com.ververica:flink-connector-mongodb-cdc:2.3.0.
+ */
+public class BsonUtils {
```
Comment (EMsnap) [inline inlong-sort/sort-connectors/mongodb-cdc/src/main/java/org/apache/inlong/sort/cdc/mongodb/source/utils/BsonUtils.java:40]: can it be reused from flinkcdc ?

Comment (e-mhui) [reply inlong-sort/sort-connectors/mongodb-cdc/src/main/java/org/apache/inlong/sort/cdc/mongodb/source/utils/BsonUtils.java]: yes, it had been removed.

**Review by dockerzhang — APPROVED** [review]

**Pull request merged**